### PR TITLE
feat: make encode/decode safe and add encodeUnsafe/decodeUnsafe

### PR DESCRIPTION
**Pull request by mikeal**
This is a change I’ve been meaning to make for a while.

Before this change, we didn’t actually cache decodes and suffered a full encode/decode in order to get a copy. We also cached the encode and exposed it directly, which is a mutable buffer so it was quite unsafe.

With this change: 

* we keep a cache of the encode/decode indefinitely
* encode/decode methods both return fully mutable copies (buffer copy for encode and deep clone for decode)
* expose encodeUnsafe() and decodeUnsafe() which return the cached version.

This means that the default encode()/decode() will be noticable slower, but if you are **sure** the result won’t be mutated you can use the `unsafe` methods and get better performance than prior versions with full caching.

This is similar to Node.js Buffer’s approach to performance and safety.

### CODE REVIEW

File: test/block.spec.js [hunk 1]
```diff
@@ -90,3 +90,14 @@ test('reader', done => {
   same(reader.get('hello').value, 'world')
   done()
 })
+
+test('decode cache', done => {
+  const block = Block.encoder({ hello: 'world' }, 'dag-cbor')
+  const decoded = block.decodeUnsafe()
+  decoded.test = true
+  assert.ok(block.decode().test)
```
Comment (vmx) [inline test/block.spec.js]: Same issue as in the "Block data caching" test.

File: test/block.spec.js [hunk 2]
```diff
@@ -18,9 +18,9 @@ test('Block encode', done => {
 
 test('Block data caching', done => {
   const block = Block.encoder({ hello: 'world' }, 'dag-cbor')
-  const encoded = block.encode()
+  const encoded = block.encodeUnsafe()
```
Comment (vmx) [inline test/block.spec.js]: Are you testing here that when you're manipulating an object that is returned by `decodeUnsafe()`, that the original object is also changed, hence that manipulation shows up in the next `decode()` step?

If yes, then I don't think we should do this. It think such a manipulation should be undefined behaviour as mentioned in the README. I don't want to see people using this fact as a feature, it is just an implementation detail.

Comment (rvagg) [reply test/block.spec.js]: I gather this test is just making assertions about caching, not that you can mutate, in which case it seems reasonable to me. An alternative might just be an identity check, `===`, which should say the same thing and doesn't demonstrate the kind of use the readme warns against.

**Comment by mikeal** [comment]
> Demonstrates the need for immutable TypedArrays

Yup, I logged an issue in `nodejs/node` for that feature while doing the initial implementation. Writing this library definitely made me feel like we don’t have enough immutable primitives in JS across the board.

File: test/block.spec.js (hunk 2 above)
Comment (mikeal) [reply test/block.spec.js]: So, you **are definitely not suppose to do this when you use the library** ;) This test actually demonstrates the hazards of using the Unsafe methods, but it’s necessary in order to test that it is actually exposing the underlying cache object and not a copy like the safe methods do.

**Pull request merged**